<a href="https://colab.research.google.com/github/smeshnayasova61/project/blob/main/%D0%A2%D0%B5%D1%81%D1%82%D0%BE%D0%B2%D0%BE%D0%B5_%D0%B7%D0%B0%D0%B4%D0%B0%D0%BD%D0%B8%D0%B5_%D0%B4%D0%BB%D1%8F_%D1%80%D0%B0%D0%B1%D0%BE%D1%82%D1%8B.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Перед вами небольшая таблица (https://clck.ru/34xJwW), содержащая информацию о покупках (cheque) в интернет-магазине в различных регионах (каждая строка - транзакция). Успешная покупка имеет status, равный 1. Необходимо проанализировать данные и ответить на вопрос:

Укажите регион, с наибольшим средним чеком покупок, совершенных за январь.

https://spb.hh.ru/applicant/vacancy_response?vacancyId=83244359

In [9]:
import pandas as pd
#import numpy as np

In [10]:
from google.colab import files
uploaded = files.upload()

Saving test - Sheet1.csv to test - Sheet1.csv


In [70]:
path = "copied path"
df = pd.read_csv('/content/test - Sheet1.csv')

In [71]:
df

date  status country_name       cheque
0      2022-12-03 00:00:00       1    Казахстан  3,221567698
1      2023-01-02 00:00:00       1     Бразилия  185,6290682
2      2023-02-17 00:00:00       1   Кыргызстан   15,8307312
3      2022-11-04 00:00:00       1       Россия  20,31545493
4      2022-12-11 00:00:00       1   Узбекистан   40,6340304
...                    ...     ...          ...          ...
20131  2022-11-10 00:00:00       1       Россия  22,40901368
20132  2023-02-20 00:00:00       1     Бразилия  171,8991052
20133  2022-12-29 00:00:00       1    Казахстан  3,490416715
20134  2022-12-30 00:00:00       1     Бразилия  168,1639455
20135  2023-02-04 00:00:00       1       Россия  17,25899893

[20136 rows x 4 columns]

In [72]:
columns = df.columns
columns

Index(['date', 'status', 'country_name', 'cheque'], dtype='object')

In [73]:
df.index

RangeIndex(start=0, stop=20136, step=1)

In [74]:
df.dtypes

date            object
status           int64
country_name    object
cheque          object
dtype: object

In [75]:
df.isna().sum()

date            0
status          0
country_name    0
cheque          0
dtype: int64

In [76]:
num_feat = list(df.select_dtypes(exclude='object').columns)
num_feat

['status']

In [77]:
#Вывод строковых переменных
obj_feat = list(df.select_dtypes(include='object').columns)
obj_feat

['date', 'country_name', 'cheque']

In [78]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20136 entries, 0 to 20135
Data columns (total 4 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   date          20136 non-null  object
 1   status        20136 non-null  int64 
 2   country_name  20136 non-null  object
 3   cheque        20136 non-null  object
dtypes: int64(1), object(3)
memory usage: 629.4+ KB


In [79]:
df["date"] = df['date'].astype('datetime64')
df.dtypes

date            datetime64[ns]
status                   int64
country_name            object
cheque                  object
dtype: object

In [80]:
df['status'].unique ()

array([1, 0])

In [81]:
df.loc[df['status'] == 1]

date  status country_name       cheque
0     2022-12-03       1    Казахстан  3,221567698
1     2023-01-02       1     Бразилия  185,6290682
2     2023-02-17       1   Кыргызстан   15,8307312
3     2022-11-04       1       Россия  20,31545493
4     2022-12-11       1   Узбекистан   40,6340304
...          ...     ...          ...          ...
20131 2022-11-10       1       Россия  22,40901368
20132 2023-02-20       1     Бразилия  171,8991052
20133 2022-12-29       1    Казахстан  3,490416715
20134 2022-12-30       1     Бразилия  168,1639455
20135 2023-02-04       1       Россия  17,25899893

[19177 rows x 4 columns]

In [82]:
date_feat = list(df.select_dtypes(include='datetime64[ns]').columns)
date_feat

['date']

In [83]:
df['month'] = pd.DatetimeIndex(df['date']).month
df['month']

0        12
1         1
2         2
3        11
4        12
         ..
20131    11
20132     2
20133    12
20134    12
20135     2
Name: month, Length: 20136, dtype: int64

In [84]:
df['month'].unique()

array([12,  1,  2, 11, 10,  3])

In [85]:
df=df.loc[df['month'] == 1]
df

date  status country_name       cheque  month
1     2023-01-02       1     Бразилия  185,6290682      1
12    2023-01-20       1     Бразилия  171,1686028      1
13    2023-01-30       1      Украина  26,40694131      1
14    2023-01-08       1       Россия  18,08627385      1
18    2023-01-16       1       Россия  19,35377171      1
...          ...     ...          ...          ...    ...
20099 2023-01-31       1       Россия  17,32748702      1
20105 2023-01-25       1       Россия   17,6317322      1
20108 2023-01-14       1        Индия  15,30033246      1
20116 2023-01-15       0      Украина   29,6853056      1
20119 2023-01-31       0     Бразилия  198,4802205      1

[4553 rows x 5 columns]

In [86]:
df['date'].unique()

array(['2023-01-02T00:00:00.000000000', '2023-01-20T00:00:00.000000000',
       '2023-01-30T00:00:00.000000000', '2023-01-08T00:00:00.000000000',
       '2023-01-16T00:00:00.000000000', '2023-01-21T00:00:00.000000000',
       '2023-01-12T00:00:00.000000000', '2023-01-10T00:00:00.000000000',
       '2023-01-05T00:00:00.000000000', '2023-01-03T00:00:00.000000000',
       '2023-01-25T00:00:00.000000000', '2023-01-19T00:00:00.000000000',
       '2023-01-13T00:00:00.000000000', '2023-01-07T00:00:00.000000000',
       '2023-01-15T00:00:00.000000000', '2023-01-23T00:00:00.000000000',
       '2023-01-28T00:00:00.000000000', '2023-01-11T00:00:00.000000000',
       '2023-01-31T00:00:00.000000000', '2023-01-14T00:00:00.000000000',
       '2023-01-24T00:00:00.000000000', '2023-01-06T00:00:00.000000000',
       '2023-01-22T00:00:00.000000000', '2023-01-29T00:00:00.000000000',
       '2023-01-01T00:00:00.000000000', '2023-01-26T00:00:00.000000000',
       '2023-01-09T00:00:00.000000000', '2023-01-18

In [87]:
df=df.loc[df['status'] == 1]
df

date  status country_name       cheque  month
1     2023-01-02       1     Бразилия  185,6290682      1
12    2023-01-20       1     Бразилия  171,1686028      1
13    2023-01-30       1      Украина  26,40694131      1
14    2023-01-08       1       Россия  18,08627385      1
18    2023-01-16       1       Россия  19,35377171      1
...          ...     ...          ...          ...    ...
20088 2023-01-14       1     Бразилия  176,6558926      1
20098 2023-01-24       1         Мали  2,148373458      1
20099 2023-01-31       1       Россия  17,32748702      1
20105 2023-01-25       1       Россия   17,6317322      1
20108 2023-01-14       1        Индия  15,30033246      1

[4353 rows x 5 columns]

In [88]:
df['status'].unique()

array([1])

In [89]:
def convert_currency(val):
    """
    Преобразует числовое значение строки в число с плавающей точкой:
     - удаляет $
     - удаляет запятые
     - преобразует в число с плавающей точкой
    """
    new_val = val.replace(',', '')
    return int(new_val)

In [90]:
df['cheque'] = df['cheque'].apply(convert_currency)
df.dtypes

<ipython-input-90-cca6b4a97ea9>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['cheque'] = df['cheque'].apply(convert_currency)


date            datetime64[ns]
status                   int64
country_name            object
cheque                   int64
month                    int64
dtype: object

In [91]:
df

date  status country_name      cheque  month
1     2023-01-02       1     Бразилия  1856290682      1
12    2023-01-20       1     Бразилия  1711686028      1
13    2023-01-30       1      Украина  2640694131      1
14    2023-01-08       1       Россия  1808627385      1
18    2023-01-16       1       Россия  1935377171      1
...          ...     ...          ...         ...    ...
20088 2023-01-14       1     Бразилия  1766558926      1
20098 2023-01-24       1         Мали  2148373458      1
20099 2023-01-31       1       Россия  1732748702      1
20105 2023-01-25       1       Россия   176317322      1
20108 2023-01-14       1        Индия  1530033246      1

[4353 rows x 5 columns]

In [92]:
df['cheque'].mean()

2099510741.9593384

In [93]:
df=df.loc[df['cheque'] > 2099510741]
df

date  status country_name      cheque  month
13    2023-01-30       1      Украина  2640694131      1
26    2023-01-12       1       Россия  2262228721      1
46    2023-01-20       1     Бразилия  2205189335      1
69    2023-01-13       1    Казахстан  3706427815      1
70    2023-01-07       1       Россия  2220117852      1
...          ...     ...          ...         ...    ...
20026 2023-01-03       1      Украина  3003810108      1
20032 2023-01-27       1    Казахстан  4000912316      1
20033 2023-01-28       1      Украина  3428660855      1
20073 2023-01-03       1    Казахстан  2970129997      1
20098 2023-01-24       1         Мали  2148373458      1

[1257 rows x 5 columns]

In [94]:
df=df.sort_values(by='cheque', ascending=False)
df

date  status country_name      cheque  month
5101  2023-01-01       1     Беларусь  9881260916      1
4670  2023-01-14       1     Беларусь  9876025528      1
1882  2023-01-01       1     Беларусь  9562063209      1
19710 2023-01-21       1      Нигерия  9556863116      1
2608  2023-01-04       1       Италия  9402693539      1
...          ...     ...          ...         ...    ...
11902 2023-01-07       1       Россия  2102880737      1
8320  2023-01-07       1       Россия  2102880737      1
3626  2023-01-29       1       Россия  2102663161      1
1224  2023-01-08       1       Россия  2102551467      1
6574  2023-01-09       1   Кот Дивуар  2099966705      1

[1257 rows x 5 columns]

In [95]:
df=df['cheque'].max()
df

9881260916


date	status	country_name	cheque	month
5101	2023-01-01	1	Беларусь	9881260916	1

Вывод: Регионом с наибольшим средним чеком покупок, совершенных за январь является Беларусь.